File: app/src/main/java/com/alex/testeandroid/data/remote/ServiceGenerator.java
package com.alex.testeandroid.data.remote;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.alex.testeandroid.BuildConfig;

import java.util.concurrent.TimeUnit;

import io.reactivex.schedulers.Schedulers;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * Created by Alex on 27/08/2018
 */
public class ServiceGenerator {

    //region FIELDS
    private static final long DEFAULT_TIMEOUT = 60;
    //endregion

    //region METHODS

    //region PUBLIC METHODS
    public static <S> S createService(Class<S> serviceClass, final String baseUrl) {
        OkHttpClient okHttpClient = new OkHttpClient.Builder()
                .connectTimeout(DEFAULT_TIMEOUT, TimeUnit.SECONDS)
                .writeTimeout(DEFAULT_TIMEOUT, TimeUnit.SECONDS)
                .readTimeout(DEFAULT_TIMEOUT, TimeUnit.SECONDS)
                .addInterceptor(getLoggingInterceptor())
                .build();

        Gson gson = new GsonBuilder()
                .create();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(baseUrl)
                .client(okHttpClient)
                .addConverterFactory(GsonConverterFactory.create(gson))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.createWithScheduler(Schedulers.io()))
                .build();

        return retrofit.create(serviceClass);
    }
    //endregion

    //region PRIVATE METHODS
    private static Interceptor getLoggingInterceptor() {
        HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
        logging.setLevel(BuildConfig.DEBUG ? HttpLoggingInterceptor.Level.BODY : HttpLoggingInterceptor.Level.NONE);
        return logging;
    }
    //endregion
    //endregion
}

File: app/src/main/java/com/rafhack/testeandroid/data/remote/ServiceGenerator.kt
package com.rafhack.testeandroid.data.remote

import com.google.gson.GsonBuilder
import com.rafhack.testeandroid.BuildConfig
import io.reactivex.schedulers.Schedulers
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.util.concurrent.TimeUnit

object ServiceGenerator {

    //region FIELDS
    private const val DEFAULT_TIMEOUT: Long = 60
    //endregion

    //region METHODS

    //region PUBLIC METHODS
    fun <S> createService(serviceClass: Class<S>): S {
        val builder = OkHttpClient.Builder()
                .connectTimeout(DEFAULT_TIMEOUT, TimeUnit.SECONDS)
                .writeTimeout(DEFAULT_TIMEOUT, TimeUnit.SECONDS)
                .readTimeout(DEFAULT_TIMEOUT, TimeUnit.SECONDS)
                .addNetworkInterceptor(defaultInterceptor)
                .addInterceptor(loggingInterceptor)

        val retrofit = Retrofit.Builder()
                .baseUrl(BuildConfig.API_URL)
                .client(builder.build())
                .addConverterFactory(GsonConverterFactory.create(GsonBuilder().create()))
                .addCallAdapterFactory(RxJava2CallAdapterFactory.createWithScheduler(Schedulers.io()))
                .build()

        return retrofit.create(serviceClass)
    }
    //endregion

    //region PRIVATE METHODS
    private val defaultInterceptor: Interceptor
        get() = Interceptor { chain ->
            val builder = chain.request().newBuilder()
            val request = builder.build()
            chain.proceed(request)
        }

    private val loggingInterceptor: Interceptor
        get() {
            val logging = HttpLoggingInterceptor()
            logging.level = if (BuildConfig.DEBUG) HttpLoggingInterceptor.Level.BODY else HttpLoggingInterceptor.Level.NONE
            return logging
        }
    //endregion

    //endregion
}